File: 05_Development/CareBridgeMobileApp/test/features/directChat/calls/direct_call_coordinator_test.dart
import 'dart:async';

import 'package:flutter_test/flutter_test.dart';
import 'package:untitled/features/directChat/calls/direct_call_api.dart';
import 'package:untitled/features/directChat/calls/direct_call_coordinator.dart';
import 'package:untitled/features/directChat/calls/direct_call_state.dart';
import 'package:untitled/features/directChat/models/conversation_call.dart';
import 'package:untitled/features/directChat/models/zego_join_credentials.dart';
import 'package:untitled/integrations/firebaseRealtime/conversation_event_signal.dart';

class _FakeCallApi implements DirectCallApiPort {
  final activeCompleter = Completer<List<ConversationCall>>();
  final detailRequests = <String>[];
  ConversationCall? detail;

  @override
  Future<List<ConversationCall>> listActiveCalls() => activeCompleter.future;

  @override
  Future<ConversationCall> getCall(String conversationId, String callId) async {
    detailRequests.add(callId);
    return detail!;
  }

  @override
  Future<ConversationCall> initiate(String conversationId, String callType) =>
      throw UnimplementedError();

  @override
  Future<ConversationCall> markRinging(
    String conversationId,
    String callId,
  ) async => detail!;

  @override
  Future<ConversationCall> answer(String conversationId, String callId) =>
      throw UnimplementedError();

  @override
  Future<ConversationCall> decline(String conversationId, String callId) =>
      throw UnimplementedError();

  @override
  Future<ConversationCall> end(String conversationId, String callId) =>
      throw UnimplementedError();

  @override
  Future<ZegoJoinCredentials> issueJoinCredentials(
    String conversationId,
    String callId,
  ) => throw UnimplementedError();
}

ConversationCall _call(String status) => ConversationCall(
  callId: 'call-1',
  conversationId: 'conversation-1',
  initiatedByUserId: 'mother-1',
  callType: 'VOICE',
  callStatus: status,
  initiatedAt: DateTime.utc(2026, 7, 16),
);

ConversationEventSignal _signal(String eventId) => ConversationEventSignal(
  eventId: eventId,
  eventType: 'CALL_STATE_CHANGED',
  conversationId: 'conversation-1',
  resourceId: 'call-1',
  occurredAt: DateTime.utc(2026, 7, 16),
);

void main() {
  test(
    'signal during initial sync is reconciled once without duplicate overlay',
    () async {
      final api = _FakeCallApi()..detail = _call('RINGING');
      final signals = StreamController<ConversationEventSignal>.broadcast();
      final coordinator = DirectCallCoordinator(
        api: api,
        signals: signals.stream,
        currentUserId: () => 'expert-1',
      );

      final start = coordinator.start();
      signals.add(_signal('event-1'));
      api.activeCompleter.complete([_call('RINGING')]);
      await start;
      await Future<void>.delayed(Duration.zero);

      expect(coordinator.state.phase, DirectCallPhase.incoming);
      expect(coordinator.state.call?.callId, 'call-1');
      expect(api.detailRequests.length, lessThanOrEqualTo(1));
      await coordinator.dispose();
      await signals.close();
    },
  );

  test(
    'duplicate and replayed signals cannot regress a terminal REST state',
    () async {
      final api = _FakeCallApi()..detail = _call('ENDED');
      final signals = StreamController<ConversationEventSignal>.broadcast();
      final coordinator = DirectCallCoordinator(
        api: api,
        signals: signals.stream,
        currentUserId: () => 'mother-1',
      );
      api.activeCompleter.complete(const []);
      await coordinator.start();

      signals
        ..add(_signal('event-1'))
        ..add(_signal('event-1'));
      await Future<void>.delayed(Duration.zero);
      await Future<void>.delayed(Duration.zero);

      expect(coordinator.state.phase, DirectCallPhase.terminal);
      expect(coordinator.state.call?.callStatus, 'ENDED');
      await coordinator.dispose();
      await signals.close();
    },
  );
}

File: 05_Development/CareBridgeWebApp/src/features/directChat/calls/directCallCoordinator.test.ts
import { describe, expect, it, vi } from 'vitest';
import type { ConversationEventSignal } from '../../../shared/integrations/firebaseRealtime/conversationEventSignal';
import type { ConversationCall, ZegoJoinCredentials } from '../models/directConversation';
import type { DirectCallApiPort } from './directCallApi';
import { DirectCallCoordinator } from './directCallCoordinator';

const call = (callStatus: string, callId = 'call-1'): ConversationCall => ({
  callId,
  conversationId: 'conversation-1',
  initiatedByUserId: 'mother-1',
  callType: 'VIDEO',
  callStatus,
  initiatedAt: '2026-07-16T00:00:00Z',
  answeredAt: callStatus === 'ANSWERED' ? '2026-07-16T00:01:00Z' : null,
  endedAt: null,
  durationSeconds: null,
});

const signal = (eventId: string, callId = 'call-1'): ConversationEventSignal => ({
  eventId,
  eventType: 'CALL_STATE_CHANGED',
  conversationId: 'conversation-1',
  resourceId: callId,
  occurredAt: '2026-07-16T00:00:00Z',
});

class FakeSignals {
  listener?: (event: ConversationEventSignal) => void;
  unsubscribe = vi.fn();

  subscribe(listener: (event: ConversationEventSignal) => void) {
    this.listener = listener;
    return this.unsubscribe;
  }
}

class FakeApi implements DirectCallApiPort {
  active = Promise.resolve<ConversationCall[]>([]);
  detail = call('RINGING');
  credentials = Promise.resolve<ZegoJoinCredentials>({
    appId: 1,
    roomId: 'cb_room',
    userId: 'u_mother',
    displayName: 'Mother',
    token: 'token',
    expiresAt: '2026-07-16T01:00:00Z',
  });
  getCall = vi.fn(async () => this.detail);
  listActiveCalls = vi.fn(() => this.active);
  markRinging = vi.fn(async () => this.detail);
  initiate = vi.fn(async () => this.detail);
  answer = vi.fn(async () => call('ANSWERED'));
  decline = vi.fn(async () => call('DECLINED'));
  end = vi.fn(async () => call('ENDED'));
  issueJoinCredentials = vi.fn(() => this.credentials);
}

describe('DirectCallCoordinator', () => {
  it('reconciles a signal arriving during initial sync once and deduplicates replay', async () => {
    let resolveActive!: (calls: ConversationCall[]) => void;
    const api = new FakeApi();
    api.active = new Promise((resolve) => {
      resolveActive = resolve;
    });
    const signals = new FakeSignals();
    const coordinator = new DirectCallCoordinator(api, signals, () => 'expert-1');

    const starting = coordinator.start();
    signals.listener?.(signal('event-1'));
    signals.listener?.(signal('event-1'));
    resolveActive([call('RINGING')]);
    await starting;
    await vi.waitFor(() => expect(coordinator.state.phase).toBe('incoming'));

    expect(api.getCall.mock.calls.length).toBeLessThanOrEqual(1);
  });

  it('ignores stale credential results after another call becomes current', async () => {
    let resolveCredentials!: (value: ZegoJoinCredentials) => void;
    const api = new FakeApi();
    api.active = Promise.resolve([call('ANSWERED')]);
    api.credentials = new Promise((resolve) => {
      resolveCredentials = resolve;
    });
    const signals = new FakeSignals();
    const coordinator = new DirectCallCoordinator(api, signals, () => 'mother-1');
    await coordinator.start();

    const pending = coordinator.issueJoinCredentialsForCurrent();
    api.detail = call('RINGING', 'call-2');
    signals.listener?.(signal('event-2', 'call-2'));
    await vi.waitFor(() => expect(coordinator.state.call?.callId).toBe('call-2'));
    resolveCredentials({
      appId: 1,
      roomId: 'cb_old',
      userId: 'u_mother',
      displayName: 'Mother',
      token: 'old-token',
      expiresAt: '2026-07-16T01:00:00Z',
    });

    await expect(pending).resolves.toBeNull();
  });

  it('unsubscribes and ignores later events on dispose', async () => {
    const api = new FakeApi();
    const signals = new FakeSignals();
    const coordinator = new DirectCallCoordinator(api, signals, () => 'mother-1');
    await coordinator.start();

    coordinator.dispose();
    signals.listener?.(signal('event-after-dispose'));
    await Promise.resolve();

    expect(signals.unsubscribe).toHaveBeenCalledOnce();
    expect(api.getCall).not.toHaveBeenCalled();
  });
});
